### ISSUE TITLE
ESQL: Imply `IS NOT NULL` constraints for some aggs/functions/operators

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#102787]
### Description

Most aggs push ignore `null` values. In this example:
```
FROM logs-postgres*
| DISSECT message \"duration: %{query_duration} ms\"
| EVAL query_duration_num = TO_DOUBLE(query_duration)
| STATS avg_duration = AVG(query_duration_num)
```

any `null` `query_duration_num`s should be

**Comment by nik9000** [comment]
Relates to #101997.

**Comment by costin** [comment]
Went after this one and implemented in #103030 only to discover that we cannot just apply the filter since it will end up the empty groups.
Consider the following entries with columns a and b
[1, x]
[2, null]

The query `stats count(a) by b` becomes `where a is not null | stats count(a) by b` which filters out group `2` as it has only null values.
This could be a dynamic filter that gets applied after discovering the set of groups (filter by filter) which makes it much less valueable.
